### DISCUSSION TITLE
Hardware instancing

### DISCUSSION THREAD

**Discussion by gibssa** [google/filament#6739] [Q&A] [answered]
Hi again Filament team,

I was hoping you could clarify [this comment](https://github.com/google/filament/pull/5199/files#diff-029a38c64fc7643973b3a8dfdecc04214f9130a62b46fe3ef0f46246147ecfe5) from your PR:

> Specifies the number of draw instance of this renderable. The default is 1 instance and
> the maximum number of instances allowed is 65535. 0 is invalid.
> All instances are culled using the same bounding box, so care must be taken to make
> sure all instances render inside the specified bounding box.
> The material can use getInstanceIndex() in the vertex shader to get the instance index and
> possibly *adjust the position or transform*.

I have an existing (repeated) transform that I'd like to batch into a single draw call. Here's my (naive) code:

```
for (int i = 0; i < 100; ++i) {
  const auto& instance = tm_.getInstance(entities_[i]);
  auto translation = mat4f::translation(float3(
      offsets.x,
      offsets.y,
      offsets.z
  ));

  float percentage = ...;
  auto scaling = mat4f::scaling(float3(
      0.020f * percentage,
      0.01f,
      0.020f * percentage
  ));
  tm_.setTransform(instance, translation * scaling);
}
```

I'm converting this to instead call `RenderableManager::Builder.instances(100)`.

Here's my (unfinished) CPU + shader code:
```
RenderableManager::Builder(1)
      .boundingBox({{0}, {1}})
      .material(0, mMaterialInstance)
      .geometry(0,
                RenderableManager::PrimitiveType::TRIANGLES,
                geometry->vertexBuffer,
                geometry->indexBuffer)
      .instances(100)
      .culling(false)
      .build(engine, mRenderable);
```
```
material {
    name : Batched,
    parameters : [
        {
           type : float4,
           name : color
        }
    ],
    blending : transparent,
    culling : none,
    instanced: true,
    depthCulling : false,
    shadingModel : unlit
}

vertex {
    // https://google.github.io/filament/Materials.html#materialdefinitions/materialblock/general:instanced
    void materialVertex(inout MaterialVertexInputs material) {
        int instance_index = getInstanceIndex();

        material.worldPosition.xyz = ??
    }
}

```

It seems that the [`material.worldPosition`](https://github.com/google/filament/blob/19ab9f0403fcd536d16e03dfa4a8146659c88bb4/shaders/src/material_inputs.vs#L78) accepts a position, but not a transform?

So my question is:
 1. How can I pass a (translation + scaling) transform to a vertex shader to batch draw calls?

Thanks!

**Comment by bejado** [comment, marked as answer]
`material.worldPosition` sets the position of a _vertex_. By default (in the OBJECT vertex domain), it is computed like so:

```
mat4 transform = getWorldFromModelMatrix();
vec3 position = getPosition().xyz;
material.worldPosition = mulMat4x4Float3(transform, position);
```

So if you want to pass a unique transform per instance, you somehow need to customize `transform` depending on the instance. How you do this depends on what you're trying to achieve. If you have some matrices on the CPU you'd like to use, you could pass them as a `float3×3` parameter array and use `getInstanceIndex()` to index into them.

**Reply by gibssa** [reply]
Thank you for the details, I'll try to implement this and close out this comment once I do.

**Reply by gibssa** [reply]
I think I got it, thanks @bejado !
